taichi/transforms/demote_operations.cpp

```diff
@@ -14,6 +14,25 @@ class DemoteOperations : public BasicStmtVisitor {
   DemoteOperations() : BasicStmtVisitor() {
   }
 
+  void visit(BitExtractStmt *stmt) override {
+    // @ti.func
+    // def bit_extract(input, begin, end):
+    //   return (input >> begin) & ((1 << (end - begin)) - 1)
+    VecStatement statements;
+    auto begin = statements.push_back<ConstStmt>(
+        LaneAttribute<TypedConstant>(stmt->bit_begin));
+    auto input_sar_begin = statements.push_back<BinaryOpStmt>(
+        BinaryOpType::bit_sar, stmt->input, begin);
+    auto mask = statements.push_back<ConstStmt>(LaneAttribute<TypedConstant>(
+        ((1 << (stmt->bit_end - stmt->bit_begin)) - 1)));
```

Comment:
nit: Please use `1LL` here because `stmt->bit_end` can be `32` while `stmt->bit_begin` is `0`. `1 << 32` overflows. However, the result should have data type the same with `input` instead of `ti.i64` (which is the default when you pass in a `long long`), so you may want to use the constructor `TypedConstant(DataType dt, const T &value)`.

Reply:
I thought we always use i32 for address?

Reply:
Yes, but it's possible that `stmt->bit_end == 32` and `stmt->bit_begin == 0` by the definition of `BitExtractStmt`. However this may never happen in the current codebase though (because we don't allow addresses of > 2^31 Bytes)...

Reply:
(And maybe we want to support i64 addresses in the future, i.e., `stmt->input->ret_type.data_type` can be i64?)